File: sdk-java/TaxBanditsAPIAuthentication/apiauthentication/src/main/java/com/span/HeaderUtils.java
package com.span;


import java.util.HashMap;

/**
 * Created by sts-184 on 11/20/17.
 */

public class HeaderUtils
{
    static final String USER_TOKEN = "YOUR USER TOKEN";

    static final String IP_ADDRESS = "YOUR IP ADDRESS";

    static final String PUBLIC_KEY = "YOUR PUBLIC KEY";

    static final String PRIVATE_KEY = "YOUR PRIVATE KEY";

    public HashMap<String, String> getHeaders(String methodName, String methodType)
    {

        HashMap<String, String> headerMap = new HashMap<String, String>();

        AuthenticationUtils authenticationUtils = new AuthenticationUtils();

        String timestamp = authenticationUtils.getDateUTCForamat();

        headerMap.put("Timestamp", timestamp);

        String authenticationKey = PUBLIC_KEY + ":" + authenticationUtils.getAuth(methodType, timestamp, IP_ADDRESS, USER_TOKEN, methodName);

        headerMap.put("Authentication", authenticationKey.trim());

        headerMap.put("UserToken", USER_TOKEN);

        headerMap.put("IpAddress", IP_ADDRESS);

        return headerMap;


    }


}

File: sdk-python/static/HeaderUtils.py
import base64
import hashlib
import hmac

from static.Utility import getCurrentUTCTime, isValidString
from static import USER_TOKEN, PUBLIC_KEY, PRIVATE_KEY, __SERVER_IP__


def getHeaders(methodType, methodName):
    """ headersUtils:
        methodType -> 'GET' or 'POST'
        methodName -> 'Demo/HelloWorld' """

    # Step 1: Add Current Date Time in UTC format eg: 'Tuesday, Nov 28, 2017 11:51:37 AM'
    utcDate = getCurrentUTCTime()

    headers = {"Timestamp": utcDate}

    # Step 2: Add Your System Ip Address
    headers['IpAddress']= __SERVER_IP__

    # Step 3: Add User token as an User Auth Key
    headers['UserToken']= USER_TOKEN

    # Step 4: Generate Authentication Key with private Key and
    message = buildAuthSignature(methodType, utcDate, __SERVER_IP__, USER_TOKEN, str.lower(methodName))

    # Step 5: Generate Authentication with public key and private key
    headers['Authentication'] = PUBLIC_KEY + ":" + computeHashKey(PRIVATE_KEY, message)

    return headers



def buildAuthSignature(methodType, utcDate, __SERVER_IP__, USER_TOKEN, absolutePath):
    """ buildAuthSignature:
            methodType   ->  'GET' or 'POST'
            date         ->  '%A, %b %d, %Y %I:%M:%S %p'  like  'Sunday, Jan 01, 2020 10:10:00 AM'
            absolutePath ->  'demo/HelloWorld'
    """

    absolutePath = "/" + absolutePath
    from urlparse import urlparse
    url=""

    if isValidString(urlparse(absolutePath).path):
        url = urlparse(absolutePath).path

    if isValidString(urlparse(absolutePath).query):
        url = url + urlparse(absolutePath).query

    sentence = [methodType, utcDate, __SERVER_IP__, USER_TOKEN, absolutePath]
    return '\n'.join(sentence)



def computeHashKey(privateKey, message):
    """ buildAuthSignature:
               privateKey   ->  Your private Key
               message      ->  buildAuthSignature generated message """

    message = bytes(message).encode('utf-8')
    secret = bytes(privateKey).encode('utf-8')
    signature = base64.b64encode(hmac.new(secret, message, digestmod=hashlib.sha256).digest())
    return signature
